src/code-considerations/criteria-for-inclusion.md

```diff
@@ -0,0 +1,70 @@
+# Criteria for inclusion in `std`
+
+**Note**: *The process for deciding what does and does not get added to the standard
+library is too complex to reduce to a set of concrete rules. Ultimately we rely
+upon the expertise of members of the library API team, but here we attempt to
+document the reasoning that has been used historically to guide contributors
+and team members and to ensure long term consistency in how these decisions are
+made even as the team membership naturally changes over time.*
+
+## Reasons For Inclusion
+
+* APIs where there is a clear best implementation
+* APIs where there is a strong benefit to having only one, such as common
+  traits (e.g. `Error`, the `Fail` trait from `failure` crate showed the cost
+  of competing error traits and how it causes ecosystem fractures)
+* Core value propositions: APIs that we want to use in examples for our
+  learning materials to demonstrate rust's core value propositions. (e.g.
+  scoped threads simplifying examples demonstrating thread and memory safety)
+* Unsafe primitives: APIs that reduce the need of or centralize difficult to
+  implement unsafe code
+* Portability: APIs that are not dependent upon any platform's abstractions or
+  features
+
```

Comment:
I think another principal is something along the lines of we want to provide the foundational building blocks to support many programming paradigms rather than choose one paradigm for all users and make it first class in std. So atomics (and mutexes to a lesser degree) are provided not because we think they are the best paradigm for concurrency but because they are required to build any paradigm.

Reply:
@nrc I don't think that's quite an accurate description. We will, sometimes, choose a common paradigm and support it in std, just not necessarily to the *exclusion* of other paradigms.

We don't just support mutexes because they're useful for building other paradigms, we also support them because they're useful in their own right *as* a paradigm.

Reply:
Historically, it has certainly been a guiding principal for how std was designed. For mutexes, I think that there is the locking paradigm there, but the interesting question is why choose that paradigm and not actors or message passing, etc.? And the answer is because it is easier to implement actors with mutexes than to implement mutexes with actors.

That philosophy may have changed recently and of course real life is more nuanced than this simple description, but I think it is a good principal and it's served us well.

Reply:
> I do also think we'd happily add support for a specific paradigm that not everyone uses if it's popular and meets all the other criteria (e.g. not TMTOWTDI).

I think this is worth discussing in some detail. IMO it begs the question why have one paradigm and not another? And it might open a slippery slope to having many paradigms in std, which seems very sub-optimal. IMO this comes back to the principal of there being only one good solution, but lifted from APIs to paradigms.